bundle/config/mutator/python/python_mutator.go

```diff
@@ -581,43 +630,79 @@ func createInitOverrideVisitor(ctx context.Context, mode insertResourceMode) mer
 			}
 
+			err := state.UpdatedResources.Add(valuePath, left)
+			if err != nil {
+				return fmt.Errorf("failed to handle path (%s): %w", valuePath.String(), err)
+			}
+
 			// deleting properties is allowed because it only changes an existing resource
 			log.Debugf(ctx, "Delete value at %q", valuePath.String())
 
 			return nil
 		},
 		VisitInsert: func(valuePath dyn.Path, right dyn.Value) (dyn.Value, error) {
-			hasResourcesPrefix := valuePath.HasPrefix(resourcesPath)
-
-			// insert 'resources' or 'resources.<resource_type>' if it didn't exist before
-			if hasResourcesPrefix && len(valuePath) <= len(resourcesPath)+1 {
-				return right, nil
-			}
-
-			if !hasResourcesPrefix {
+			if !valuePath.HasPrefix(resourcesPath) {
 				return dyn.InvalidValue, fmt.Errorf("unexpected change at %q (insert)", valuePath.String())
 			}
 
-			// valuePath has 3 elements: 'resources.<resource_type>.<resource_name>'
-			insertResource := len(valuePath) == len(resourcesPath)+2
-			if mode == insertResourceModeDisallow && insertResource {
-				return dyn.InvalidValue, fmt.Errorf("unexpected change at %q (insert)", valuePath.String())
+			// for len = 3: insert 'resources.<resource_type>.<resource_name>'
+			// for len = 2: insert 'resources.<resource_type>' (added the first resource of type)
+			// for len = 1: insert 'resources' (added the first resource)
+			insertResource := len(valuePath) <= len(resourcesPath)+2
+			if insertResource {
+				if mode == insertResourceModeDisallow {
+					return dyn.InvalidValue, fmt.Errorf("unexpected change at %q (insert)", valuePath.String())
+				}
+
+				err := state.AddedResources.Add(valuePath, right)
+				if err != nil {
+					return dyn.InvalidValue, fmt.Errorf("failed to handle path (%s): %w", valuePath.String(), err)
+				}
+			} else {
+				err := state.UpdatedResources.Add(valuePath, right)
+				if err != nil {
+					return dyn.InvalidValue, fmt.Errorf("failed to handle path (%s): %w", valuePath.String(), err)
+				}
 			}
 
 			log.Debugf(ctx, "Insert value at %q", valuePath.String())
 
 			return right, nil
 		},
 		VisitUpdate: func(valuePath dyn.Path, left, right dyn.Value) (dyn.Value, error) {
-			if !valuePath.HasPrefix(jobsPath) {
+			if !valuePath.HasPrefix(resourcesPath) {
 				return dyn.InvalidValue, fmt.Errorf("unexpected change at %q (update)", valuePath.String())
 			}
 
+			// update is similar to insert, the only difference is presence of left (previous) value
+
+			// for len = 3: insert 'resources.<resource_type>.<resource_name>'
+			// for len = 2: insert 'resources.<resource_type>' (added the first resource of type)
+			// for len = 1: insert 'resources' (added the first resource)
+			insertResource := len(valuePath) <= len(resourcesPath)+2
+			if insertResource {
+				if mode == insertResourceModeDisallow {
+					return dyn.InvalidValue, fmt.Errorf("unexpected change at %q (update)", valuePath.String())
+				}
+
+				err := state.AddedResources.Add(valuePath, right)
```

Comment:
So visit is called for every node but we only care about first 3 levels, right? resources.<type>.<name>

Can we just read the keys of those maps explicitly and avoid extra work (as well as Add() function have to support extra cases).